### PR TITLE
fix: Fix Stream not close because it holds in TreeOfLoser

### PR DESCRIPTION
**Pull request by jinchengchenghh**
SpillPartion::createOrderedReader opens all the spill files and holds it in TreeOfLoser, SpillFile -> FileInputStream->buffer_(readBufferSize), so it holds memory readBufferSize * numFiles,  so we need to close the stream as soon as possible after it reaches the end as Spark does. 
After close, the function SpillMergeStream::hasData() returns false, then it won't access the fields of SpillMergeStream. 
https://github.com/apache/spark/blob/master/core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java#L134

### CODE REVIEW

File: velox/exec/Spill.h [hunk 1]
```diff
@@ -172,7 +179,10 @@ class FileSpillBatchStream : public BatchStream {
   }
 
   bool nextBatch(RowVectorPtr& batch) override {
-    return spillFile_->nextBatch(batch);
+    if (spillFile_->nextBatch(batch)) {
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:182]: why make the change? It is the same as
```
return spillFile_->nextBatch(batch);
```

File: velox/exec/Spill.h [hunk 2]
```diff
@@ -125,6 +127,9 @@ class SpillMergeStream : public MergeStream {
 
   // Covers all rows inn 'rowVector_' Set if 'decoded_' is non-empty.
   SelectivityVector rows_;
+
+  // True if the stream is closed.
+  bool closed_;
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:132]: When closed_ is set?
```
bool closed_{false}
```
Move to first? And shall we add to check in API that closed_ is not set?

File: velox/exec/Spill.cpp [hunk 3]
```diff
@@ -78,6 +78,14 @@ int32_t SpillMergeStream::compare(const MergeStream& other) const {
   return 0;
 }
 
+void SpillMergeStream::close() {
+  rowVector_.reset();
```
Comment (xiaoxmeng) [inline velox/exec/Spill.cpp:82]: VELOX_CHECK(!closed_);
closed_ = true;

**Review by xiaoxmeng — COMMENTED** [review]
@jinchengchenghh good change % comments. Thanks!

File: velox/exec/Spill.h (hunk 1 above)
Comment (jinchengchenghh) [reply velox/exec/Spill.h:182]: I make a change before, dropped close() in the final version. I think we don't need to close the `FileSpillBatchStream`. Back it.

File: velox/exec/Spill.cpp (hunk 3 above)
Comment (jinchengchenghh) [reply velox/exec/Spill.cpp:86]: Thanks for catching, fixed it.

**Comment by jinchengchenghh** [comment]
The read CI LocalRunnerTest cased by https://github.com/facebookincubator/velox/issues/11619

File: velox/exec/Spill.h [hunk 4]
```diff
@@ -147,15 +156,19 @@ class FileSpillMergeStream : public SpillMergeStream {
   }
 
   int32_t numSortKeys() const override {
+    VELOX_CHECK(!closed_);
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:159]: No need to check in this method

File: velox/exec/Spill.h [hunk 5]
```diff
@@ -62,6 +64,7 @@ class SpillMergeStream : public MergeStream {
   /// batch, in which case the caller must call copy out current batch data if
   /// required before calling pop().
   vector_size_t currentIndex(bool* isLastRow = nullptr) const {
+    VELOX_CHECK(!closed_);
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:67]: ditto

File: velox/exec/Spill.h [hunk 6]
```diff
@@ -54,6 +55,7 @@ class SpillMergeStream : public MergeStream {
   void pop();
 
   const RowVector& current() const {
+    VELOX_CHECK(!closed_);
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:58]: Let's keep this

File: velox/exec/Spill.h [hunk 7]
```diff
@@ -147,15 +156,19 @@ class FileSpillMergeStream : public SpillMergeStream {
   }
 
   int32_t numSortKeys() const override {
+    VELOX_CHECK(!closed_);
     return spillFile_->numSortKeys();
   }
 
   const std::vector<CompareFlags>& sortCompareFlags() const override {
+    VELOX_CHECK(!closed_);
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:164]: ditto

File: velox/exec/Spill.h [hunk 8]
```diff
@@ -46,6 +46,7 @@ class SpillMergeStream : public MergeStream {
   }
 
   bool operator<(const MergeStream& other) const final {
+    VELOX_CHECK(!closed_);
```
Comment (xiaoxmeng) [inline velox/exec/Spill.h:49]: no need here as we check inside compare? thanks!

**Review by xiaoxmeng — APPROVED** [review]
@jinchengchenghh thanks for the update. LGTM % nits.

File: velox/exec/Spill.h (hunk 4 above)
Comment (jinchengchenghh) [reply velox/exec/Spill.h:159]: We get the sortKeys and compareFlags from spillFile_, after close, the spillFile is reset, so I check it here.

File: velox/exec/Spill.h (hunk 5 above)
Comment (jinchengchenghh) [reply velox/exec/Spill.h:67]: Same reason. reset the `rowVector_` in close function.

**Pull request closed without merging**